Run template validate spec present

waciumawanjohi opened the pull request:
<!--
Thanks for submitting a pull request to Cartographer!

Also check the [PR guidelines] if you haven't already!
-->

[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

<!--
Add the story that is resolved or updated
`closes`/`fixes` or `updates` are valid here 
-->
Should RunTemplates accept objects that do not have a spec? If an object does not have a `spec` field, there will be a cache miss when Carto examines the object. This will cause a new object to be stamped out.
Consider #587 

<!--
Summarize your changes. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->


## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Add CreatorName field in Workload spec

Example notes:

* Operators can label stamped objects with the Workload's creator
* Creators can be contacted when interdepartmental issues arise 

If there are no additional notes necessary you may remove this entire section.
-->
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [x] Removed non-atomic or `wip` commits
- ~[ ] Filled in the [Release Note](#Release-Note) section above~
- ~[ ] Modified the docs to match changes~ <!-- TBD: reference doc editing guidance -->

cirocosta commented:
> _If an object does not have a spec field, there will be a cache miss when Carto examines the object. This will cause a new object to be stamped out._

I wonder if that means we should try to address this at the cacher rather than individually at the Runnable level (where we can perform inspections like this because there's no `ytt` support) 🤔 

I'm not very familiar with the internals of the caching layer - in the case of a supplychain that stamps out a ConfigMap, what's the current impact of those cache misses?

waciumawanjohi commented:
>  in the case of a supplychain that stamps out a ConfigMap, what's the current impact of those cache misses?

For a supply chain template, if it deploys a resource that doesn't have a spec, the template will always update the object. Observable behavior would be generation increasing with every reconcile loop.

To fix this at the caching layer (which I think it desirable), we need to have some fields that we expect to be set by cartographer and other fields that we expect to be set by the object's reconciler. (for example: `spec` is Carto's concern and `status` is the other controller's concern). Right now our line in the sand is: Carto is concerned with spec and all other fields are the concern of the other controller. Looking at configmaps, `data` would be a reasonable addition to Carto's concern. There's an exploration that would need to be done to validate the fields that can be assumed to be Carto's concern.

waciumawanjohi committed:
ClusterRunTemplate accepts only objects with a spec

Improve the UX by adding a reasonable constraint.
Runnables assume that the templated object has a spec field.

squeedee committed:
Merge branch 'main' into run-template-validate-spec-present
